src/providers/ipfs.ts

```diff
@@ -70,43 +72,97 @@ export class PinJob extends Job {
+    logger.verbose(`(${hash}) Retrieving meta size of CID`)
+    const metadataSizeMb = await this.getMetaFileSize(cid) // In MB
 
-      if (metadataSize.gt(this.expectedSize)) {
-        logger.error(`The hash ${hash} has cumulative size of ${bytesToMegabytes(stats.CumulativeSize)} megabytes while it was expected to have ${this.expectedSize} megabytes.`)
-        throw new HashExceedsSizeError('The hash exceeds payed size!', bytesToMegabytes(stats.CumulativeSize), this.expectedSize)
-      }
-    } catch (e) {
-      if (e.name === 'TimeoutError') {
-        logger.error(`Fetching size of ${hash} timed out!`)
-        throw new Error(`Fetching size of ${hash} timed out!`)
-      } else {
-        throw e
-      }
+    if (metadataSizeMb.gt(this.expectedSize)) {
+      logger.error(`The hash ${hash} has cumulative size of ${metadataSizeMb.toString()} megabytes while it was expected to have ${this.expectedSize} megabytes.`)
+      throw new HashExceedsSizeError('The hash exceeds payed size!', metadataSizeMb, this.expectedSize)
     }
 
-    await this.swarmConnect().catch(logger.warn)
-
-    logger.info(`Pinning hash: ${hash} start`)
+    // PIN PROCESS
     // We can be generous on the actual pinning timeout as if the CID would not be present
     // in IPFS network, then the previous ipfs.object.stat() call would timeout already then.
     // We are using 0.5 MB per second transfer rate, with keeping at least 20 minutes as default.
     // SizeInMB * 0.5 * 1000 ==> ms
-    const estimatedTimeout = Math.max(MIN_PIN_TIMEOUT, metadataSize.div(RATE_MB_PER_SECOND).multipliedBy(1000).toNumber())
-    await this.ipfs.pin.add(cid, { timeout: estimatedTimeout })
+    const estimatedTimeout = Math.max(MIN_PIN_TIMEOUT, metadataSizeMb.div(RATE_MB_PER_SECOND).multipliedBy(1000).toNumber())
+    await this.pinProcess(cid, { timeout: estimatedTimeout })
 
-    await this.swarmDisconnect().catch(logger.warn)
+    // ACTUAL SIZE CHECK
+    logger.verbose(`(${hash}) Retrieving actual size of CID`)
+    const sizeInMb = await this.getActualFileSize(cid) // In MB
+
+    if (sizeInMb.gt(this.expectedSize)) {
+      logger.error(`The hash ${hash} has cumulative size of ${sizeInMb.toString()} megabytes while it was expected to have ${this.expectedSize} megabytes.`)
+      // Unpin file
+      logger.info(`Unpin file ${hash} due to exceed size limit`)
+      await this.ipfs.pin.rm(cid)
+      throw new HashExceedsSizeError('The hash exceeds payed size!', sizeInMb, this.expectedSize)
+    }
   }
 }
 
+export function getDagStat (nodeUrl: string): (cid: CID, options?: any) => Promise<any> {
+  return (cid: CID, options?: RequestInit): Promise<any> =>
+    fetch(`${nodeUrl}/dag/stat?arg=${cid.toString()}`, { method: 'POST', ...options })
+      .then(async res => {
+        if (!res.ok) {
+          throw new Error(`Get dag stat for hash ${cid.toString()} error, ${res.statusText}`)
+        }
+        return (await res.text())
+          .split('\n')
```

Comment:
I am not convinced that this custom parsing is needed 😳
When I do:

```
$ curl -X POST "http://127.0.0.1:5001/api/v0/dag/stat?arg=QmUszxPhL5s7LHyn7ehbFTu9DmQUbbdKniTpJkQr7EpZgR&progress=true"
```

I get normal looking JSON:

```
{"Size":51,"NumBlocks":1}
```

What are you getting?

Reply:
![image](https://user-images.githubusercontent.com/41945483/100211498-e1dadb00-2f14-11eb-9dd8-387b41afe7fe.png)
This what I am talking about

Reply:
Hmm I believe that is because of the `progress=true` ;-) We don't need progress here, so just don't use the flag...

Reply:
I see I had it in the CURL command there as well, only my data were small enough that it did not emit the progress... 

Try it without it, please :-)

Reply:
Same result

Reply:
![image](https://user-images.githubusercontent.com/41945483/100212623-421e4c80-2f16-11eb-90d0-c0de6db4a3a4.png)

Reply:
Hmmm so it seems progress is true by default :-D So use `progress=false` ;-)

Reply:
Ahh, yes. Thanks 👍